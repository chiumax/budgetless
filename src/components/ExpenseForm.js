import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import swal from "sweetalert";
import NavigationPrompt from "react-router-navigation-prompt";
import { BigNumber } from "bignumber.js";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense
        ? BigNumber(props.expense.amount)
            .div(100)
            .toString(10)
        : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
      onFormDirty: false,
      modalTitle: "Are You Sure?",
      modalText:
        "Once you leave the page, you will not be able to recover your edits.",
      modalIcon: "warning",
      modalButtons: {
        confirm: { text: "Leave", value: true },
        cancel: "Stay"
      }
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    const formdirty = e.target.value.length > 0;
    this.setState(() => ({
      description,
      onFormDirty: formdirty
    }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    const formdirty = e.target.value.length > 0;

    this.setState(() => ({
      note,
      onFormDirty: formdirty
    }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    const formdirty = e.target.value.length > 0;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount, onFormDirty: formdirty }));
    }
  };

  onDateChange = createdAt => {
    const formdirty = true;

    if (createdAt) {
      this.setState(() => ({ createdAt, onFormDirty: formdirty }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      //   Set error state equal to 'Please provide description and amount.'
      this.setState(() => ({
        error: "Please provide a description and an amount value."
      }));
    } else {
      //   Clear the error
      const modalTemp = this.props.expense
        ? "Expense Edited!"
        : "Expense Created!";
      this.setState(() => ({
        error: "",
        onFormDirty: false,
        modalTitle: modalTemp,
        modalText:
          "You can edit/view expenses by clicking on them at the Dashboard",
        modalIcon: "success",
        modalButtons: {
          confirm: { text: "Ok", value: true }
        }
      }));
      if (this.props.expenseLength === true) {
        this.setState(() => ({
          modalIcon: "error",
          modalButtons: {
            confirm: { text: "Ok", value: true }
          },
          modalText:
            "We're sorry for limiting this account to 5 expenses since we are currently hosted using a free service. ",
          modalTitle: "We're sorry."
        }));
      }
      this.props.onSubmit({
        description: this.state.description,
        amount: BigNumber(this.state.amount)
          .multipliedBy(100)
          .toString(10),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  componentWillUnmount() {}
  hideModal = () => {
    swal.close();
  };
  render() {
    return (
      <div>
        <NavigationPrompt when={this.state.onFormDirty}>
          {({ isActive, onConfirm, onCancel }) => {
            if (isActive) {
              swal({
                title: this.state.modalTitle,
                text: this.state.modalText,
                icon: this.state.modalIcon,
                buttons: this.state.modalButtons,
                dangerMode: true,
                closeOnEsc: false,
                closeOnClickOutside: false
              }).then(willDelete => {
                if (willDelete) {
                  onConfirm();
                } else {
                  onCancel();
                  this.hideModal();
                }
              });
            }
          }}
        </NavigationPrompt>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
            maxLength="50"
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
            maxLength="20"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense (optional)"
            maxLength="500"
          />
          <button>{this.props.expense ? "Edit Expense" : "Add Expense"}</button>
        </form>
      </div>
    );
  }
}
