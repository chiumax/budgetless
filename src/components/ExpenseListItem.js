import React from "react";
import { NavLink } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import { BigNumber } from "bignumber.js";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <NavLink to={"/edit/" + id}>
      <h3>{description}</h3>
    </NavLink>
    <p>{id}</p>
    <p>
      {numeral(BigNumber(amount).div(100)).format("$0,0.00")}-
      {moment(createdAt).format("MMMM Do, YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;
