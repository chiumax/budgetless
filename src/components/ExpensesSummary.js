import React from "react";
import { connect } from "react-redux";
import getExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";
import { BigNumber } from "bignumber.js";

export const ExpensesSummary = props => (
  <div>
    <p>
      Currently viewing {props.expenseCount}{" "}
      {props.expenseCount > 1 ? "expenses" : "expense"} totalling{" "}
      {numeral(BigNumber(props.expensesTotal).div(100)).format("$0,0.00")}
    </p>
  </div>
);

const mapStateToProps = state => {
  return {
    expenseCount: state.expenses.length,
    expensesTotal: getExpensesTotal(state.expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
