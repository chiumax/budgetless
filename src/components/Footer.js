import React from "react";
import { Emoji } from "emoji-mart";

const Footer = () => (
  <div className="content-container">
    <h4>
      Made with <Emoji emoji="heart" native={true} size={16} /> by{" "}
      <a href="https://github.com/dumblole" className="link-hover">
        dumblole
      </a>
    </h4>
  </div>
);

export default Footer;
