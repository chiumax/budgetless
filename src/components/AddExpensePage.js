import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import Anime from "react-anime";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    if (this.props.expenseLength === false) {
      this.props.startAddExpense(expense);
    }
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <Anime
          easing="easeInExpo"
          duration={1000}
          direction="normal"
          translateX={["-13rem", 0]}
          opacity={[0, 1]}
        >
          <div className="page-header">
            <div className="content-container">
              <h1 className="page-header__title">Add An Expense</h1>
            </div>
          </div>
          <div className="content-container">
            <ExpenseForm
              onSubmit={this.onSubmit}
              expenseLength={this.props.expenseLength}
            />
          </div>
        </Anime>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

const mapStateToProps = state => {
  return {
    expenseLength: state.expenses.length >= 5
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddExpensePage);
