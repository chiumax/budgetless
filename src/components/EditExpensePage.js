import React from "react";
import { connect } from "react-redux";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
import Anime from "react-anime";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    if (this.props.expenseLength === false) {
      this.props.startEditExpense(this.props.expense.id, expense);
    }

    this.props.history.push("/");
  };
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <Anime
          easing="easeInExpo"
          duration={1000}
          direction="normal"
          translateX={["-13rem", 0]}
          opacity={[0, 1]}
        >
          <div className="page-header">
            <div className="content-container">
              <h1 className="page-header__title">Edit Expense</h1>
            </div>
          </div>
          <div className="content-container">
            <ExpenseForm
              expense={this.props.expense}
              onSubmit={this.onSubmit}
              expenseLength={false}
            />
            <button className="button button--secondary" onClick={this.onClick}>
              Remove Expense
            </button>
          </div>
        </Anime>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
