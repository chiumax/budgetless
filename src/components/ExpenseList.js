import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div>
    Some text to make this easy to identify on the screen.
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(item => {
        return <ExpenseListItem key={item.id} {...item} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
