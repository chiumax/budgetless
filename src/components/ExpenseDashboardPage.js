import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";
import Anime from "react-anime";

const ExpenseDashboardPage = () => (
  <div>
    <Anime
      easing="easeInExpo"
      duration={1000}
      direction="normal"
      delay={(el, index) => index * 240}
      translateX={["-13rem", 0]}
      opacity={[0, 1]}
    >
      <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
      </div>
    </Anime>
  </div>
);

export default ExpenseDashboardPage;
