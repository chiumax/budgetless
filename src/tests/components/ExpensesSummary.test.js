import { ExpensesSummary } from "../../components/ExpensesSummary";
import { shallow } from "enzyme";
import React from "react";

test("test the expenseCount for how many visible expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={23} expensesTotal={"22222"} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("test the total amount of the expenses for the variable expensesTotal", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expensesTotal={"123"} />
  );
  expect(wrapper).toMatchSnapshot();
});
