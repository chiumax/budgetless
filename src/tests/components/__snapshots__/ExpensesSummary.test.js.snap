// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the expenseCount for how many visible expenses 1`] = `
<div>
  <p>
    Currently viewing 
    23
     
    expenses
     totalling
     
    $222.22
  </p>
</div>
`;

exports[`test the total amount of the expenses for the variable expensesTotal 1`] = `
<div>
  <p>
    Currently viewing 
    1
     
    expense
     totalling
     
    $1.23
  </p>
</div>
`;
